scripts/autotracking.lua

```diff
@@ -295,35 +295,56 @@ function handleMoonstone(keyItem)
       moonstone.CurrentStage = 0
     end
   end
-  
+
 end
 
 --
 -- Handle items that can also show up in character's equipment slots.
--- This includes the Hero Medal and Robo's Ribbon. Masamune is handled
--- separately since it is more complicated.
+-- This includes the Hero Medal, Robo's Ribbon, and Grand Leon.
+-- Masamune is handled separately based on reforging with Melchior.
 --
 function handleEquippableItem(keyItem)
 
-  local equipmentSlot = AutoTracker:ReadU8(keyItem.address, 0)
-  local itemOwned = keyItem.found or equipmentSlot == keyItem.value
-  
-  local trackerItem = Tracker:FindObjectForCode(keyItem.name)  
-  if trackerItem and not trackerItem.Owner.ModifiedByUser then
-    trackerItem.Active = itemOwned
+  local trackerItem = Tracker:FindObjectForCode(keyItem.name)
+
+  if trackerItem.Owner.ModifiedByUser then
+    return
+  end
+
+  if not trackerItem then
+    trackerItem.Active = false
+    return
+  end
+
+  if keyItem.found then
+    trackerItem.Active = true
+    return
   end
 
+  -- to support Duplicate characters, need to check each character
+  local found = false
+  for pc=0,5 do
+    local address = 0x7E2650 + 0x50*pc + keyItem.offset
+    local equipmentSlot = AutoTracker:ReadU8(address, 0)
```

Comment:
It might be better to add one (or several) memory watches than to use the AutoTracker:Read* functions.  As I understand it, these will block while doing a round trip to the emulator/hardware to read the values from memory, so doing this multiple times in succession may cause a performance hit.  Of course this also introduces weirdness where you have to track both the inventory and character equip slots in separate memory reads that are firing async.  


Random side note:
There is an issue with EmoTracker when using the bundled connector.lua where AutoTracker:Read* will sometime flake out and always return zeroes until autotracking is stopped and restarted.  The workaround is to use SNI or qusb2snes (rather than the bundled CrowdControl based scritps), so I don't think this is an issue in Poptracker.

Reply:
Makes sense. Thanks for feedback.

So something like the `updateItemsFromInventory` approach, I'm thinking? Can do :+1:

Reply:
Oh, I see what you mean about "weirdness", get a bit of a race condition if there are two memory watches for inventory and equipment.

Looks like inventory is 0x7E2400 + 0xF2 and Equipment is 0x7E2600 + 0x25A. If that was combined into one watch, it's like 1114, which is like twice as big as the "Events" watch. So that's probably pretty big...

Reply:
Sadly, qusb2snes is only 1024. Maybe add a watch just on the character data we need, then add 2 more fields to the equippable items in the map - equipped and inventory?  Then we can just or those two flags together in both callbacks to set the toggle state of the item?

Alternately, if you aren't seeing any issues we can just keep using the immediate read for now. Then if it becomes an issue in the future we can change to a memory watch.

Reply:
Haha, shucks, it looks like to cover all inventory plus all character slots it's 1034, so close!

Yeah, I think having separate `.found` and `.equipped` fields to set makes sense. Going to play with that.

Thanks!

Reply:
Cool made update from https://github.com/coffeemancy/Jets-of-Time-Tracker/pull/5